File: cache_interceptor/src/main/java/uk/co/glass_software/android/cache_interceptor/interceptors/cache/CacheManager.java
package uk.co.glass_software.android.cache_interceptor.interceptors.cache;

import android.support.annotation.NonNull;

import com.google.gson.Gson;

import java.util.Date;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import uk.co.glass_software.android.cache_interceptor.response.ResponseMetadata;
import uk.co.glass_software.android.cache_interceptor.utils.Action;
import uk.co.glass_software.android.cache_interceptor.utils.Function;
import uk.co.glass_software.android.cache_interceptor.utils.Logger;

import static uk.co.glass_software.android.cache_interceptor.interceptors.cache.CacheToken.Status.CACHED;
import static uk.co.glass_software.android.cache_interceptor.interceptors.cache.CacheToken.Status.STALE;

class CacheManager {
    
    private final Function<Long, Date> dateFactory;
    private final Gson gson;
    private final Logger logger;
    private final long timeToLiveInMs;
    private final DatabaseManager databaseManager;
    
    CacheManager(DatabaseManager databaseManager,
                 Function<Long, Date> dateFactory,
                 Gson gson,
                 Logger logger,
                 int timeToLiveInMinutes) {
        this.databaseManager = databaseManager;
        this.dateFactory = dateFactory;
        this.gson = gson;
        this.logger = logger;
        this.timeToLiveInMs = timeToLiveInMinutes * 60000L;
    }
    
    public void clearOlderEntries() {
        //FIXME
//        runAsync(databaseManager::clearOlderEntries);
    }
    
    public void flushCache() {
        runAsync(databaseManager::flushCache);
    }
    
    private void runAsync(Action action) {
        Completable.fromAction(action::act)
                   .subscribeOn(Schedulers.io())
                   .subscribe();
    }
    
    @SuppressWarnings("unchecked")
    <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> Observable<R> getCachedResponse(
            Observable<R> upstream,
            @NonNull CacheToken<R> cacheToken) {
        String simpleName = cacheToken.getResponseClass().getSimpleName();
        logger.d(this, "Checking for cached " + simpleName);
        
        R cachedResponse = databaseManager.getCachedResponse(upstream, cacheToken);
        
        if (cachedResponse == null) {
            logger.d(this, "No cached " + simpleName);
            return fetchAndCache(upstream, cacheToken);
        }
        else {
            ResponseMetadata<R, E> metadata = cachedResponse.getMetadata();
            CacheToken<R> cachedResponseToken = metadata.getCacheToken();
            CacheToken.Status status = getCachedStatus(cachedResponseToken);
            metadata.setCacheToken(CacheToken.newStatus(cachedResponseToken, status));
            
            logger.d(this, "Found cached " + simpleName + ", status: " + status);
            
            if (status == STALE) {
                return refreshStale(cachedResponse, upstream);
            }
            else {
                return Observable.just(cachedResponse);
            }
        }
    }
    
    @NonNull
    <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> CacheToken.Status getCachedStatus(
            @NonNull CacheToken<? extends R> cacheToken) {
        Date expiryDate = cacheToken.getExpiryDate();
        if (expiryDate == null) {
            return STALE;
        }
        return dateFactory.get(null).getTime() > expiryDate.getTime() ? STALE : CACHED;
    }
    
    private <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> Observable<R> fetchAndCache(
            Observable<R> upstream,
            CacheToken<R> cacheToken) {
        String simpleName = cacheToken.getResponseClass().getSimpleName();
        logger.d(this, "Fetching and caching new " + simpleName);
        
        return upstream
                .doOnNext(response -> {
                    logger.d(this, "Finished fetching " + simpleName + ", now delivering");
                    ResponseMetadata<R, E> metadata = response.getMetadata();
                    if (metadata.getError() == null) {
                        Date fetchDate = dateFactory.get(null);
                        Date expiryDate = dateFactory.get(fetchDate.getTime() + timeToLiveInMs);
                        metadata.setCacheToken(CacheToken.caching(cacheToken,
                                                                  upstream,
                                                                  fetchDate,
                                                                  fetchDate,
                                                                  expiryDate
                        ));
                    }
                })
                .doAfterNext(response -> {
                    if (response.getMetadata().getError() == null) {
                        logger.d(this, simpleName + " successfully delivered, now caching");
                        databaseManager.cache(response);
                    }
                });
    }
    
    @SuppressWarnings("unchecked")
    private <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> Observable<R> refreshStale(
            R cachedResponse,
            Observable<R> upstream) {
        String simpleName = cachedResponse.getClass().getSimpleName();
        ResponseMetadata<R, E> metadata = cachedResponse.getMetadata();
        CacheToken<R> cacheToken = metadata.getCacheToken();
        
        logger.d(this, simpleName + " is " + cacheToken.getStatus() + ", attempting to refresh");
        
        Observable<R> fetchAndCache = fetchAndCache(upstream, cacheToken)
                .map(response -> {
                    E error = response.getMetadata().getError();
                    
                    boolean isCouldNotRefresh = error != null
                                                && response.getMetadata().getError().get(error);
                    
                    if (isCouldNotRefresh) {
                        return deepCopy(cachedResponse, CacheToken.Status.COULD_NOT_REFRESH);
                    }
                    else {
                        return updateRefreshed(response, CacheToken.Status.REFRESHED);
                    }
                });
        
        return Observable.just(Observable.just(cachedResponse), fetchAndCache)
                         .concatMap(observable -> observable.subscribeOn(Schedulers.io()));
    }
    
    @SuppressWarnings("unchecked")
    private <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> R deepCopy(
            R cachedResponse,
            CacheToken.Status newStatus) {
        R copiedResponse = gson.fromJson(gson.toJson(cachedResponse),
                                         (Class<R>) cachedResponse.getClass()
        );
        ResponseMetadata<R, E> metadata = cachedResponse.getMetadata();
        CacheToken newToken = CacheToken.newStatus(metadata.getCacheToken(), newStatus);
        metadata.setCacheToken(newToken);
        return copiedResponse;
    }
    
    private <E extends Exception & Function<E, Boolean>, R extends ResponseMetadata.Holder<R, E>> R updateRefreshed(
            R response,
            CacheToken.Status status) {
        String simpleName = response.getClass().getSimpleName();
        ResponseMetadata<R, E> metadata = response.getMetadata();
        CacheToken<R> newToken = CacheToken.newStatus(metadata.getCacheToken(), status);
        metadata.setCacheToken(newToken);
        
        logger.d(this,
                 "Delivering "
                 + (status == CacheToken.Status.REFRESHED ? "refreshed" : "stale")
                 + " "
                 + simpleName
                 + ", status: "
                 + newToken.getStatus()
        );
        
        return response;
    }
}

File: cache_interceptor/src/main/java/uk/co/glass_software/android/cache_interceptor/interceptors/cache/CacheManager.kt
package uk.co.glass_software.android.cache_interceptor.interceptors.cache

import androidx.annotation.VisibleForTesting
import com.google.gson.Gson
import io.reactivex.Completable
import io.reactivex.Observable
import uk.co.glass_software.android.boilerplate.utils.log.Logger
import uk.co.glass_software.android.boilerplate.utils.rx.On
import uk.co.glass_software.android.boilerplate.utils.rx.schedule
import uk.co.glass_software.android.cache_interceptor.annotations.CacheInstruction
import uk.co.glass_software.android.cache_interceptor.interceptors.cache.CacheStatus.*
import uk.co.glass_software.android.cache_interceptor.response.CacheMetadata
import uk.co.glass_software.android.cache_interceptor.response.ResponseWrapper
import java.util.*


internal class CacheManager<E>(private val databaseManager: DatabaseManager<E>,
                               private val dateFactory: (Long?) -> Date,
                               private val gson: Gson,
                               private val logger: Logger)
        where E : Exception,
              E : (E) -> kotlin.Boolean {

    fun clearCache(typeToClear: Class<*>?, //TODO
                   clearOlderEntriesOnly: Boolean) =
            Completable.create {
                if (clearOlderEntriesOnly) {
                    databaseManager.clearOlderEntries()
                } else {
                    databaseManager.clearCache()
                }
                it.onComplete()
            }!!

    fun getCachedResponse(upstream: Observable<ResponseWrapper<E>>,
                          instructionToken: CacheToken,
                          cacheOperation: CacheInstruction.Operation.Expiring,
                          isNetworkError: (E) -> Boolean)
            : Observable<ResponseWrapper<E>> {

        val instruction = instructionToken.instruction
        val simpleName = instruction.responseClass.simpleName
        logger.d("Checking for cached $simpleName")

        val cachedResponse = databaseManager.getCachedResponse(instructionToken)

        if (cachedResponse == null) {
            logger.d("No cached $simpleName")
            return fetchAndCache(
                    upstream,
                    cacheOperation,
                    instructionToken
            )
        } else {
            val metadata = cachedResponse.metadata!!
            val cachedResponseToken = metadata.cacheToken
            val status = getCachedStatus(cachedResponseToken)
            cachedResponse.metadata = metadata.copy(
                    cacheToken = cachedResponseToken.copy(status = status)
            )
            logger.d("Found cached $simpleName, status: $status")

            return if (status === STALE) {
                refreshStale(
                        cachedResponse,
                        cacheOperation,
                        upstream,
                        isNetworkError
                )
            } else {
                Observable.just(cachedResponse)
            }
        }
    }

    @VisibleForTesting
    fun getCachedStatus(cacheToken: CacheToken) = cacheToken.expiryDate?.let {
        if (dateFactory(null).time > it.time) STALE else CACHED
    } ?: STALE

    private fun fetchAndCache(upstream: Observable<ResponseWrapper<E>>,
                              cacheOperation: CacheInstruction.Operation.Expiring,
                              instructionToken: CacheToken)
            : Observable<ResponseWrapper<E>> {

        val simpleName = instructionToken.instruction.responseClass.simpleName
        logger.d("Fetching and caching new $simpleName")

        return upstream
                .doOnNext { responseWrapper ->
                    logger.d("Finished fetching $simpleName, now delivering")
                    val metadata = responseWrapper.metadata!!

                    if (metadata.exception == null) {
                        val fetchDate = dateFactory(null)
                        val timeToLiveInMs = cacheOperation.durationInMillis
                        val expiryDate = dateFactory(fetchDate.time + timeToLiveInMs.toLong())

                        responseWrapper.metadata = metadata.copy(
                                cacheToken = CacheToken.caching(
                                        instructionToken,
                                        fetchDate,
                                        fetchDate,
                                        expiryDate
                                )
                        )
                    }
                }
                .doAfterNext { response ->
                    if (response.metadata!!.exception == null) {
                        logger.d("$simpleName successfully delivered, now caching")
                        databaseManager.cache(
                                instructionToken,
                                cacheOperation,
                                response
                        ).subscribe({}, { logger.e(it, "Could not cache $simpleName") })
                    }
                }
    }

    private fun refreshStale(cachedResponse: ResponseWrapper<E>,
                             refreshOperation: CacheInstruction.Operation.Expiring,
                             upstream: Observable<ResponseWrapper<E>>,
                             isNetworkError: (E) -> Boolean)
            : Observable<ResponseWrapper<E>> {

        val metadata = cachedResponse.metadata!!
        val cacheToken = metadata.cacheToken
        val simpleName = cacheToken.instruction.responseClass.simpleName

        logger.d("$simpleName is ${cacheToken.status}, attempting to refresh")

        val fetchAndCache = fetchAndCache(
                upstream,
                refreshOperation,
                cacheToken
        ).map { responseWrapper ->
            val error = responseWrapper.metadata!!.exception
            val isCouldNotRefresh = error != null && isNetworkError(error)

            if (isCouldNotRefresh) {
                cachedResponse.copy(
                        response = cachedResponse.response,
                        metadata = updateStatus(cachedResponse, COULD_NOT_REFRESH)
                )
            } else {
                updateRefreshed(responseWrapper, REFRESHED)
            }
        }

        return Observable.just(Observable.just(cachedResponse), fetchAndCache)
                .concatMap { observable -> observable.schedule(On.Io, On.Trampoline) }
                as Observable<ResponseWrapper<E>>
    }

    private fun updateRefreshed(response: ResponseWrapper<E>,
                                newStatus: CacheStatus): ResponseWrapper<E> {
        val simpleName = response.metadata!!.cacheToken.instruction.responseClass.simpleName
        response.metadata = updateStatus(response, newStatus)

        logger.d("Delivering ${if (newStatus === REFRESHED) "refreshed" else "stale"} $simpleName, status: $newStatus")

        return response
    }

    private fun updateStatus(response: ResponseWrapper<E>,
                             newStatus: CacheStatus): CacheMetadata<E> {
        val metadata = response.metadata
        val oldToken = metadata!!.cacheToken
        val newToken = oldToken.copy(status = newStatus)

        return metadata.copy(cacheToken = newToken)
    }

}